python/ray/_private/worker.py

```diff
@@ -1905,9 +1905,24 @@ def shutdown(_exiting_interpreter: bool = False):
     _shutdown_all_compiled_dags()
 
     if _exiting_interpreter and global_worker.mode == SCRIPT_MODE:
-        # This is a duration to sleep before shutting down everything in order
-        # to make sure that log messages finish printing.
-        time.sleep(0.5)
+        # Checks whether all log messages for this job have been forwarded to stdout
+        # by the `ray_print_logs` thread. If not, waits for 0.5 seconds and retries
+        # the check up to 10 times, ensuring that excessive logs have time to be
+        # printed before shutting down.
+        wait_times = 10
+        wait_interval = 0.5
+        total_wait_time = wait_times * wait_interval
+        for _ in range(wait_times):
+            time.sleep(wait_interval)
+            if global_worker.gcs_log_subscriber.is_empty:
```

Comment:
We need to check `if hasattr(worker, "gcs_log_subscriber"):`?

Reply:
Fixed in https://github.com/ray-project/ray/pull/49337/commits/42e290ffc09df49f276683cc0688237747751db2